### ISSUE TITLE
Extract username from basic auth and forward it to backends

### ISSUE THREAD

**Issue by zhaohuabing** [envoyproxy/gateway#2947]
I would ask for an update of the basic auth in the `SecurityPolicy` to extract the username and add it to the headers sent to the backends.

The use case is the following, after user auth success I need to add the username in the request header that will be used by the application after. 

This depends on the Envoy upstream PR: https://github.com/envoyproxy/envoy/pull/32943

**Comment by akorosi** [comment]
+1

**Comment by s0uky** [comment]
+1 
Is there some progress in this issue, pls? 
Thank you

**Comment by s0uky** [comment]
not stale

**Comment by surenraju** [comment]
I noticed that `BasicAuthPerRoute` does not support the `forward_username_header` field—only `BasicAuth` does.

Reference: [basic_auth.proto#L56](https://github.com/envoyproxy/envoy/blob/main/api/envoy/extensions/filters/http/basic_auth/v3/basic_auth.proto#L56)

```
envoy-proxy  | [2025-02-18 18:14:36.611][1][info][main] [source/server/server.cc:1060] exiting
envoy-proxy  | Unable to unpack as envoy.extensions.filters.http.basic_auth.v3.BasicAuthPerRoute: [type.googleapis.com/envoy.extensions.filters.http.basic_auth.v3.BasicAuth] {
envoy-proxy  |   users {
envoy-proxy  |     inline_string: "suren:{SHA}ABw9tVKwslVgEf5YCxffU9+LB+CM= "
envoy-proxy  |   }
envoy-proxy  |   forward_username_header: "x-custom-user"
envoy-proxy  | }
```

From my understanding, when a `SecurityPolicy` is attached to an `HTTPRoute`, it creates an Envoy route configuration using `BasicAuthPerRoute`. For example, the generated route configuration looks like this:

https://gateway.envoyproxy.io/docs/tasks/security/basic-auth/

```
"routes": [
  {
    "match": {
      "prefix": "/"
    },
    "route": {
      "cluster": "httproute/default/backend/rule/0",
      "upgrade_configs": [
        {
          "upgrade_type": "websocket"
        }
      ]
    },
    "metadata": {
      "filter_metadata": {
        "envoy-gateway": {
          "resources": [
            {
              "kind": "HTTPRoute",
              "namespace": "default",
              "name": "backend"
            }
          ]
        }
      }
    },
    "typed_per_filter_config": {
      "envoy.filters.http.basic_auth": {
        "@type": "type.googleapis.com/envoy.extensions.filters.http.basic_auth.v3.BasicAuthPerRoute",
        "users": {
          "inline_bytes": "W3JlZGFjdGVkXQ=="
        }
      }
    },
    "name": "httproute/default/backend/rule/0/match/0/www_example_com"
  }
]
```

Would it be possible to support `forward_username_header` with `SecurityPolicy` at `HTTPRoute`?

**Comment by arkodg** [comment]
my bad @surenraju , just realized what you were referring to, EG uses `extensions.filters.http.basic_auth.v3.BasicAuthPerRoute`
https://github.com/envoyproxy/gateway/blob/fe22aca10f6709094742d4b6ae77469e7044db2e/internal/xds/translator/basicauth.go#L150C6-L150C29 for per route attachment which doesnt support `forward_username_header`
@zhaohuabing any recommendation on how to solve this one ?
should we enhance `BasicAuthPerRoute` in upstream or change the implementation to use listener filters for both cases ?

**Comment by zhaohuabing** [comment]
> I noticed that BasicAuthPerRoute does not support the forward_username_header field—only BasicAuth does.

@s0uky @arkodg We can use the generic approach to support per-rout configuration: create one BasicAuth filter in HCM filter chain per route, and enable it on the route.

You can find an example here:  https://github.com/envoyproxy/gateway/blob/main/internal/xds/translator/oidc.go

**Comment by zhaohuabing** [comment]
By the way, if the use case is to pass the username for auth purpose, using metadata is a better approach since it doesn't leak the username to the backend services.

username in metadata isn't supported yet in envoy, but implementing this feature would be a trivial change.

**Comment by arkodg** [comment]
@s0uky @surenraju @akorosi is your use case to perform authorization of these users in EG, or simply pass the user name to the backend ?

**Comment by surenraju** [comment]
@zhaohuabing, @arkodg  if we have two `HTTPRoute` resources, `backend-1 `and `backend-2`, and the user wants to send the `user` header only to` backend-1` and **not** to `backend-2`, this cannot be achieved (to my understanding). This limitation arises because we cannot override the `forward_username_header` in `BasicAuthPerRoute`, and we use the first matching route found in its HCM filter chain.

https://github.com/envoyproxy/gateway/blob/main/internal/xds/translator/basicauth.go#L61

`forward_username_header` can be enabled for all routes however this behavior can cause confusion.

**Comment by surenraju** [comment]
understand now; we should create one entry per route instead of the current approach, which uses the first matching route found in the HCM filter chain.

**Comment by zhaohuabing** [comment]
>  understand now; we should create one entry per route instead of the current approach, which uses the first matching route found in the HCM filter chain.

Yes, we need to change the current BasicAuth implementation where only one BasicAuth filter is created on the HCM filter chain. Instead, creating a separate BasicAuth filter for each HTTPRoute so they can have their own 'forward_username_header' setting.
